File: backend/src/domain/customers/entities/Customer.ts
import { PhoneNumber } from '../value-objects/PhoneNumber';

export class Customer {
  constructor(
    public readonly id: string,
    public readonly phone: PhoneNumber,
    public readonly fullName: string,
    public readonly tenantId: string,
    public readonly address?: string,
    public readonly notes?: string,
    public readonly city?: string,
    public readonly isVip: boolean = false,
    public readonly loyaltyPoints: number = 0,
    public readonly isActive: boolean = true,
    public readonly createdAt: Date = new Date(),
    public readonly updatedAt: Date = new Date()
  ) {}

  static create(
    id: string,
    phone: PhoneNumber,
    fullName: string,
    tenantId: string,
    address?: string,
    notes?: string,
    city?: string
  ): Customer {
    return new Customer(
      id,
      phone,
      fullName,
      tenantId,
      address,
      notes,
      city,
      false,
      0,
      true,
      new Date(),
      new Date()
    );
  }

  addLoyaltyPoints(points: number): Customer {
    return new Customer(
      this.id,
      this.phone,
      this.fullName,
      this.tenantId,
      this.address,
      this.notes,
      this.city,
      this.isVip,
      this.loyaltyPoints + points,
      this.isActive,
      this.createdAt,
      new Date()
    );
  }

  setVipStatus(isVip: boolean): Customer {
    return new Customer(
      this.id,
      this.phone,
      this.fullName,
      this.tenantId,
      this.address,
      this.notes,
      this.city,
      isVip,
      this.loyaltyPoints,
      this.isActive,
      this.createdAt,
      new Date()
    );
  }

  deactivate(): Customer {
    return new Customer(
      this.id,
      this.phone,
      this.fullName,
      this.tenantId,
      this.address,
      this.notes,
      this.city,
      this.isVip,
      this.loyaltyPoints,
      false,
      this.createdAt,
      new Date()
    );
  }

  update(
    fullName?: string,
    phone?: PhoneNumber,
    address?: string,
    notes?: string,
    city?: string,
    isVip?: boolean
  ): Customer {
    return new Customer(
      this.id,
      phone || this.phone,
      fullName || this.fullName,
      this.tenantId,
      address !== undefined ? address : this.address,
      notes !== undefined ? notes : this.notes,
      city !== undefined ? city : this.city,
      isVip !== undefined ? isVip : this.isVip,
      this.loyaltyPoints,
      this.isActive,
      this.createdAt,
      new Date()
    );
  }
}

File: owner_app/lib/models/customer.dart
class Customer {
  final String id;
  final String fullName;
  final String phone;
  final String? address;
  final String? city;
  final String? notes;
  final bool isActive;
  final DateTime? createdAt;

  Customer({
    required this.id,
    required this.fullName,
    required this.phone,
    this.address,
    this.city,
    this.notes,
    this.isActive = true,
    this.createdAt,
  });

  factory Customer.fromJson(Map<String, dynamic> json) {
    return Customer(
      id: json['id']?.toString() ?? '',
      fullName: json['fullName']?.toString() ?? '',
      phone: json['phone']?.toString() ?? '',
      address: json['address']?.toString(),
      city: json['city']?.toString(),
      notes: json['notes']?.toString(),
      isActive: json['isActive'] ?? true,
      createdAt: json['createdAt'] != null ? DateTime.tryParse(json['createdAt'].toString()) : null,
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'fullName': fullName,
      'phone': phone,
      if (address != null && address!.isNotEmpty) 'address': address,
      if (city != null && city!.isNotEmpty) 'city': city,
      if (notes != null && notes!.isNotEmpty) 'notes': notes,
      'isActive': isActive,
    };
  }

  Customer copyWith({
    String? id,
    String? fullName,
    String? phone,
    String? address,
    String? city,
    String? notes,
    bool? isActive,
    DateTime? createdAt,
  }) {
    return Customer(
      id: id ?? this.id,
      fullName: fullName ?? this.fullName,
      phone: phone ?? this.phone,
      address: address ?? this.address,
      city: city ?? this.city,
      notes: notes ?? this.notes,
      isActive: isActive ?? this.isActive,
      createdAt: createdAt ?? this.createdAt,
    );
  }
}
